### PR TITLE
[MPA] Recover Semi-SL for multi-class classification in OTX

### PR DESCRIPTION
**Pull request by JihwanEom**
### Summary
- Enable Semi-SL for multi-class classification
- Separate SemiClsHead into SemiLinearClsHead/SemiNonLinearClsHead
- Refactor SemiSLClassifier to minimize overriding, aligned with updated mmcls

Related PR: https://github.com/openvinotoolkit/training_extensions/pull/1415

### CODE REVIEW

File: mpa/cls/trainer.py [hunk 1]
```diff
@@ -65,8 +65,12 @@ def run(self, model_cfg, model_ckpt, data_cfg, **kwargs):
                     dash_line)
 
         # Data
-        if 'unlabeled' in cfg.data:
+        if 'unlabeled' in cfg.data and cfg.train_type == 'SEMISUPERVISED':
             datasets = [[build_dataset(cfg.data.train), build_dataset(cfg.data.unlabeled)]]
+            datasets[0][0].samples_per_gpu = cfg.data.samples_per_gpu
+            datasets[0][0].workers_per_gpu = cfg.data.workers_per_gpu
+            datasets[0][1].samples_per_gpu = cfg.data.unlabeled.get('samples_per_gpu', cfg.data.samples_per_gpu*3)
```
Comment (harimkang) [inline mpa/cls/trainer.py]: This is likely to cause index errors. As far as I know, there may be no unlabeled dataset.

File: mpa/cls/__init__.py [hunk 2]
```diff
@@ -37,7 +37,8 @@
 import mpa.modules.models.heads.custom_hierarchical_linear_cls_head
 import mpa.modules.models.heads.custom_hierarchical_non_linear_cls_head
 
-import mpa.modules.models.heads.semisl_cls_head
+import mpa.modules.models.heads.semi_linear_cls_head
+import mpa.modules.models.heads.semi_non_linear_cls_head
```
Comment (harimkang) [inline mpa/cls/__init__.py:43]: I don't know why it has to be separated into two. Are the results different depending on the head? Or do not support it? As far as I know, `semisl_cls_head` can support all models.

File: mpa/cls/trainer.py [hunk 3]
```diff
@@ -161,8 +165,8 @@ def train_worker(gpu, dataset, cfg, distributed, validate, timestamp, meta):
                 sub_loaders = [
                     build_dataloader(
                         sub_ds,
-                        sub_ds.samples_per_gpu if hasattr(sub_ds, 'samples_per_gpu') else cfg.data.samples_per_gpu,
-                        sub_ds.workers_per_gpu if hasattr(sub_ds, 'workers_per_gpu') else cfg.data.workers_per_gpu,
+                        sub_ds.samples_per_gpu,
```
Comment (harimkang) [inline mpa/cls/trainer.py]: Will all tests pass even if the attr check disappears?

File: recipes/stages/classification/semisl.yaml [hunk 4]
```diff
@@ -20,3 +21,6 @@ custom_hooks:
   priority: ABOVE_NORMAL
 - type: SemiSLClsHook
 
+task_adapt:
```
Comment (harimkang) [inline recipes/stages/classification/semisl.yaml:24]: This seems like something worth discussing.

**Review by harimkang — CHANGES_REQUESTED** [review]
I left a comment because I wasn't sure why the head had to be split into two, and other than that, I left a comment on the part that might cause an error.

File: mpa/cls/trainer.py (hunk 1 above)
Comment (harimkang) [reply mpa/cls/trainer.py]: + It looks good to be able to receive the unlabeled batch size from otx. (not a fixed 3x value)
It seems inconvenient to have to edit this code to fix this (unlabeled batch size).

File: mpa/cls/__init__.py (hunk 2 above)
Comment (JihwanEom) [reply mpa/cls/__init__.py:43]: Yes, `semisl_cls_head` can support all models but if then, for mobilenetv3, there is architecture difference (we can't load model weights from supervised/incremental). To prevent that, I've separated into `semisl_linear_cls_head` and `semisl_non_linear_cls_head`.

File: recipes/stages/classification/semisl.yaml (hunk 4 above)
Comment (JihwanEom) [reply recipes/stages/classification/semisl.yaml:24]: I agree, please share your nice thinks @jaegukhyun , @supersoob , @goodsong81

File: mpa/cls/__init__.py (hunk 2 above)
Comment (harimkang) [reply mpa/cls/__init__.py:43]: Does the head have a big impact on the load model? Also, will this change the results? For reference, all of the previous Semi-SL experiments were the results of one head.

File: mpa/cls/trainer.py (hunk 3 above)
Comment (JihwanEom) [reply mpa/cls/trainer.py]: For supervised/incremental learning, there was no problem. I'll try to catch edge cases for these lines.

File: mpa/cls/__init__.py (hunk 2 above)
Comment (JihwanEom) [reply mpa/cls/__init__.py:43]: Yes, significantly affects. (92.84 -> 0.04) when inference with supervised template. I think the output of semi-sl should support inference with supervised template.

File: mpa/cls/trainer.py (hunk 1 above)
Comment (JihwanEom) [reply mpa/cls/trainer.py]: Yes, please share your ideas if you have good suggestions. Same mechanism can be applied on other tasks, I think. @supersoob, @jaegukhyun

File: mpa/cls/__init__.py (hunk 2 above)
Comment (harimkang) [reply mpa/cls/__init__.py:43]: Well, that's right. If that's the case, please tie those two heads together into one file. semi_sl_head.py

Comment (JihwanEom) [reply mpa/cls/__init__.py:43]: Hmm, but other heads are separated, and I followed the convention. Do you think existing files also should be merged in one file?

Comment (harimkang) [reply mpa/cls/__init__.py:43]: Don't worry about the other files for now.

Comment (JihwanEom) [reply mpa/cls/__init__.py:43]: Okay, I'll reflect your comments. Thank you.

File: mpa/cls/trainer.py (hunk 1 above)
Comment (jaegukhyun) [reply mpa/cls/trainer.py]: We should discuss about the way to build unlabeled dataset, in detection case we use unlabeled data hook

File: mpa/modules/datasets/composed_dataloader.py [hunk 5]
```diff
@@ -46,6 +46,7 @@ def set_epoch(self, epoch):
     def __init__(self, loaders=[]):
         self.loaders = loaders
         self.max_iter = len(self.loaders[0])
+        self.dataset = [i.dataset for i in self.loaders]
```
Comment (jaegukhyun) [inline mpa/modules/datasets/composed_dataloader.py]: Maybe self.datasets is more propriate name?

File: mpa/cls/trainer.py (hunk 1 above)
Comment (harimkang) [reply mpa/cls/trainer.py]: I think it is necessary to expose the unlabeled batch size in the otx template. Other tasks may be run later with different settings (different unlabeled batch size).

File: recipes/stages/_base_/data/cls_semi_data.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+_base_ = [
```
Comment (jaegukhyun) [inline recipes/stages/_base_/data/cls_semi_data.py:1]: Should it be moved to otx/algo/cls/model/data_pipeline.py?

File: recipes/stages/classification/semisl.yaml (hunk 4 above)
Comment (jaegukhyun) [reply recipes/stages/classification/semisl.yaml:24]: I don't get what is the point of discussing. If it is about adding task_adapt flag to semisl recipe, we agreed on semi-sl should support task_adapt and ignore label.

File: mpa/modules/datasets/composed_dataloader.py (hunk 5 above)
Comment (JihwanEom) [reply mpa/modules/datasets/composed_dataloader.py]: I agree, but please refer ema hook. If we set the attribute as self.datasets on ComposedDL, that can be occur some.. complex if statement in ema hook.

File: recipes/stages/_base_/data/cls_semi_data.py (hunk 6 above)
Comment (JihwanEom) [reply recipes/stages/_base_/data/cls_semi_data.py:1]: Supervised/incremental is not separated with data_pipeline.py and model.py. I think we need anothet PR to align implementation method for Semi-SL include unlabeled_hook in classification. (also for segmentation)

File: recipes/stages/classification/semisl.yaml (hunk 4 above)
Comment (harimkang) [reply recipes/stages/classification/semisl.yaml:24]: As far as I know it is most likely not supported in semi right away. In particular, ignore labels are not supported due to differences in loss.

**Review by jaegukhyun — CHANGES_REQUESTED** [review]
Generally it looks good to me, I left some comments. And I have two question
1. In classification task, does configure function in mpa/cls/stage.py works same in semi-sl?
2. What is the difference between in linear head and non-linear head? Does it is number of layers of head(linear: 1, non-linear: multiple)?

File: recipes/stages/classification/semisl.yaml (hunk 4 above)
Comment (JihwanEom) [reply recipes/stages/classification/semisl.yaml:24]: Yes, I think we can resolve this conversation as-is.

File: recipes/stages/_base_/data/cls_semi_data.py (hunk 6 above)
Comment (harimkang) [reply recipes/stages/_base_/data/cls_semi_data.py:1]: In my opinion, unlabeled_hook is not needed for classification.

File: recipes/stages/classification/semisl.yaml (hunk 4 above)
Comment (jaegukhyun) [reply recipes/stages/classification/semisl.yaml:24]: Okay, then @JihwanEom do you have plan to solve task adapt issue and ignore label issue in semi-sl cls in this PR or in this sprint?

File: mpa/modules/datasets/composed_dataloader.py (hunk 5 above)
Comment (jaegukhyun) [reply mpa/modules/datasets/composed_dataloader.py]: Got it

File: mpa/cls/trainer.py (hunk 1 above)
Comment (supersoob) [reply mpa/cls/trainer.py]: Agree with @harimkang and @jaegukhyun. I think we should find a way to adapt batch size for unlabeled dataset. And I also used unlabeled data hook for seg. I hope you go over this part. I don't think it doesn't seem good to load composedDL by checking the dataset length.

File: mpa/modules/datasets/composed_dataloader.py (hunk 5 above)
Comment (supersoob) [inline mpa/modules/datasets/composed_dataloader.py]: i.dataset is weird a little. could you rename i to loader..?

File: mpa/modules/hooks/model_ema_v2_hook.py [hunk 7]
```diff
@@ -41,7 +41,10 @@ def __init__(self, ema_decay=0.9995, interval=1, start_epoch=0, dataset_len_thr=
 
     def before_train_epoch(self, runner):
         if not hasattr(self, "use_ema"):
-            self.use_ema = len(runner.data_loader.dataset) > self.dataset_len_thr
+            if isinstance(runner.data_loader.dataset, list):
```
Comment (eunwoosh) [inline mpa/modules/hooks/model_ema_v2_hook.py]: are there any chances that runner.data_loader.dataset has list type?

Comment (jaegukhyun) [reply mpa/modules/hooks/model_ema_v2_hook.py]: Jihwan add dataset attr as a list for ComposedDL in this PR

File: recipes/stages/_base_/data/cls_semi_data.py (hunk 6 above)
Comment (supersoob) [reply recipes/stages/_base_/data/cls_semi_data.py:1]: I suggest to align it by setting up the meeting. Could you summarize the agendas? I will upload PR till the meeting, so we can compare the code change. I personally think that data_pipeline doesn't have to be separated when it globally used across the models and unlabeled_hook is needed.

File: mpa/modules/hooks/model_ema_v2_hook.py (hunk 7 above)
Comment (supersoob) [inline mpa/modules/hooks/model_ema_v2_hook.py]: Checking up if it's semi or not by type is quite strange for me. I hope we can revisit this too.

File: recipes/stages/_base_/data/cls_semi_data.py (hunk 6 above)
Comment (jaegukhyun) [reply recipes/stages/_base_/data/cls_semi_data.py:1]: @JihwanEom Sorry, for my mistake. I don't know that classification does not have default data_pipeline.py in supervised learning
@harimkang I don't want to impose use of unlabeled data hook for building unlabeled dataset and ComposeDL. However I want to unify the way to build unlabeled dataset and ComposedDL. What do you think of this?

**Review by supersoob — CHANGES_REQUESTED** [review]
I cannot see stage.py change here for separation from incremental config. I doubt that it can cause the side effect -> loss/sampler use.. I hope you check this again.

File: recipes/stages/classification/semisl.yaml (hunk 4 above)
Comment (JihwanEom) [reply recipes/stages/classification/semisl.yaml:24]: No, I'm not mainly responsible, only focusing on recovering semi-sl in otx. @sovrasov might be responsible for future works in multi-class semi sl.

File: mpa/modules/hooks/model_ema_hook.py [hunk 8]
```diff
@@ -125,6 +125,15 @@ def _copy_model(self):
 
     def _ema_model(self):
         momentum = min(self.momentum, 1.0)
+
```
Comment (supersoob) [inline mpa/modules/hooks/model_ema_hook.py]: Please remove if it's not used anymore.

File: mpa/cls/trainer.py (hunk 1 above)
Comment (JihwanEom) [reply mpa/cls/trainer.py]: These code lines are removed, please check updated code to support unlabeled batch size. Thank you.

File: mpa/cls/__init__.py (hunk 2 above)
Comment (JihwanEom) [reply mpa/cls/__init__.py:43]: Can I postpone merging semi-linear head and semi-nonlinear head? I felt we need a design discussion about this.

File: mpa/cls/trainer.py (hunk 3 above)
Comment (JihwanEom) [reply mpa/cls/trainer.py]: These lines are also updated, please review again. Thank you.

File: mpa/modules/datasets/composed_dataloader.py (hunk 5 above)
Comment (JihwanEom) [reply mpa/modules/datasets/composed_dataloader.py]: Removed, thank you for suggestion.

File: mpa/modules/hooks/model_ema_v2_hook.py (hunk 7 above)
Comment (JihwanEom) [reply mpa/modules/hooks/model_ema_v2_hook.py]: Also removed by using unlabeled hook, thank you for suggestion.

**Comment by JihwanEom** [comment]
Close by https://github.com/openvinotoolkit/training_extensions/pull/1473
